state/etcd/etcd.go

```diff
@@ -0,0 +1,481 @@
+	}
+
+	reqVal, err := e.marshal(req.Value)
+	if err != nil {
+		return err
+	}
+
+	keyWithPath := e.keyPrefixPath + "/" + req.Key
+	return e.doSet(ctx, keyWithPath, reqVal, req.ETag, ttlInSeconds)
+}
+
+func (e *Etcd) BulkSet(ctx context.Context, req []state.SetRequest) error {
+	if len(req) == 0 {
+		return nil
+	}
+
+	ttls := make([]int, 0)
+	for i := 0; i < len(req); i++ {
+		ttlInSeconds, err := e.doSetValidateParameters(&req[i])
+		if err != nil {
+			return err
+		}
+		ttls = append(ttls, ttlInSeconds)
+	}
+
+	etags := make([]*string, 0)
+	for _, dr := range req {
+		keyWithPath := e.keyPrefixPath + "/" + dr.Key
+		err := e.doValidateEtag(keyWithPath, dr.ETag, dr.Options.Concurrency)
+		if err != nil {
+			return err
+		}
+		etags = append(etags, dr.ETag)
+	}
+
+	for i, dr := range req {
+		reqVal, err := e.marshal(dr.Value)
+		if err != nil {
+			return err
+		}
+
+		keyWithPath := e.keyPrefixPath + "/" + dr.Key
+		err = e.doSet(ctx, keyWithPath, reqVal, etags[i], ttls[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *Etcd) marshal(v interface{}) (string, error) {
+	var reqVal string
+	switch obj := v.(type) {
+	case []byte:
+		reqVal = string(obj)
+	case string:
+		reqVal = fmt.Sprintf("%s", obj)
+	default:
+		return "", fmt.Errorf("request value %v is not valid", reqVal)
+	}
+	return reqVal, nil
+}
+
+func (e *Etcd) doSet(ctx context.Context, key, reqVal string, etag *string, ttlInSeconds int) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if ttlInSeconds > 0 {
+		resp, err := e.client.Grant(ctx, int64(ttlInSeconds))
+		if err != nil {
+			return fmt.Errorf("couldn't grant lease %s: %s", key, err)
+		}
+		if etag != nil {
+			etag, _ := strconv.ParseInt(*etag, 10, 64)
+			_, err = e.client.Txn(ctx).
+				If(clientv3.Compare(clientv3.ModRevision(key), "=", etag)).
+				Then(clientv3.OpPut(key, reqVal, clientv3.WithLease(resp.ID))).
+				Commit()
+		} else {
+			_, err = e.client.Put(ctx, key, reqVal, clientv3.WithLease(resp.ID))
+		}
+		if err != nil {
+			return fmt.Errorf("couldn't set key %s: %s", key, err)
+		}
+	} else {
+		var err error
+		if etag != nil {
+			etag, _ := strconv.ParseInt(*etag, 10, 64)
+			_, err = e.client.Txn(ctx).
+				If(clientv3.Compare(clientv3.ModRevision(key), "=", etag)).
+				Then(clientv3.OpPut(key, reqVal)).
+				Commit()
+		} else {
+			_, err = e.client.Put(ctx, key, reqVal)
+		}
+		if err != nil {
+			return fmt.Errorf("couldn't set key %s: %s", key, err)
+		}
+	}
+	return nil
+}
+
+func (e *Etcd) doSetValidateParameters(req *state.SetRequest) (int, error) {
+	err := state.CheckRequestOptions(req.Options)
+	if err != nil {
+		return 0, err
+	}
+
+	ttlInSeconds, err := doParseTTLInSeconds(req.Metadata)
+	if err != nil {
+		return 0, err
+	}
+
+	return ttlInSeconds, nil
+}
+
+func doParseTTLInSeconds(metadata map[string]string) (int, error) {
+	s := metadata["ttlInSeconds"]
+	if s == "" {
+		return 0, nil
+	}
+
+	i, err := strconv.Atoi(s)
```

Comment:
If you use `strconv.ParseInt`, you get an `int64` already, which cane be used without type casting from `int` to `int64`